push_test/pushclient.py

```diff
@@ -0,0 +1,335 @@
+        """Send a websocket "hello" message
+
+        :param uaid: User Agent ID (if reconnecting)
+
+        """
+        if not self.connection or self.connection.state != 1:
+            await self.cmd_connect()
+        log_msg(out=self.output, status="Sending Hello")
+        args = dict(messageType="hello", use_webpush=1, **kwargs)
+        if uaid:
+            args['uaid'] = uaid
+        elif self.uaid:
+            args['uaid'] = self.uaid
+        await self._send(**args)
+
+    async def cmd_ack(self, channelID=None, version=None,
+                      timeout=60, **kwargs):
+        """Acknowledge a previous mesage
+
+        :param channelID: Channel to acknowledge
+        :param version: Version string for message to acknowledge
+        :param kwargs: Additional optional arguments
+        :param timeout: Time to wait for notifications (used by testing)
+        :return:
+
+        """
+        timeout = timeout * 2
+        while not len(self.notifications):
+            log_msg(out=self.output,
+                    status="No notifications recv'd, Sleeping...")
+            await asyncio.sleep(0.5)
+            timeout -= 1
+            if timeout < 1:
+                raise PushException("Timeout waiting for messages")
+        try:
+            while True:
+                notif = self.notifications.pop(0)
+                log_msg(out=self.output,
+                        status="Sending ACK",
+                        channelID=channelID or notif['channelID'],
+                        version=version or notif['version'])
+                await self._send(messageType="ack",
+                                 channelID=channelID or notif['channelID'],
+                                 version=version or notif['version'],
+                                 no_recv=True)
+        except IndexError:
+            pass
+        cmd, args = self._next_task()
+        await getattr(self, cmd)(**args)
+
+    async def cmd_register(self, channelID=None, key=None, **kwargs):
+        """Register a new ChannelID
+
+        :param channelID: UUID for the channel to register
+        :param key: applicationServerKey for a restricted access channel
+        :param kwargs: additional optional arguments
+        :return:
+
+        """
+        log_msg(out=self.output,
+                status="Sending new channel registration")
+        channelID = channelID or self.channelID or str(uuid.uuid4())
+        args = dict(messageType='register',
+                    channelID=channelID)
+        if key:
+            args[key] = key
+        args.update(kwargs)
+        await self._send(**args)
+
+    async def cmd_done(self, **kwargs):
+        """Close all connections and mark as done
+
+        :param kwargs: ignored
+        :return:
+
+        """
+        log_msg(out=self.output,
+                status="done")
+        await self.cmd_close()
+        await self.connection.close_connection()
```

Comment:
It feels a little odd that cmd_close also calls self.connection.close() which apparently is not enough, because here we also call close_connection().